Дать рекомендацию

s3nkyh opened the pull request.

Commit:
added: RecomContrl, RecomDto, RecomService, SkillOfferDto

Commit:
бааааааааааааааааааааааааги

Commit:
реализовал механику добавления рекомендации

Commit:
удалил recommendationId из SkillOfferDto

Commit:
поправил RecommendationController

SergeyErmoshenko commented:
Коммит с багами лучше удалить из PR

SergeyErmoshenko commented on src/main/java/school/faang/user_service/dto/recommendation/RecommendationDto.java:
```diff
@@ -0,0 +1,20 @@
+@AllArgsConstructor
+@Getter
+@Setter
+public class RecommendationDto {
```
DTO лучше делать как record. Так будет меньше кода и читать будет легче

SergeyErmoshenko commented on src/main/java/school/faang/user_service/controller/RecommendationController.java:
```diff
@@ -0,0 +1,18 @@
+    private final RecommendationService recommendationService;
+
+    public RecommendationDto giveRecommendation(RecommendationDto recommendation) {
+        ValidationUtils.validateRecommendation(recommendation);
```
Перенеси валидацию в service, так как это бизнес логика

SergeyErmoshenko commented on src/main/java/school/faang/user_service/utils/ValidationUtils.java:
```diff
@@ -0,0 +1,40 @@
+
+public class ValidationUtils {
+    public static void validateRecommendation(RecommendationDto recommendationDto) {
+        if (recommendationDto.getContent().isBlank()) {
```
А если content null?

SergeyErmoshenko commented on src/main/java/school/faang/user_service/utils/ValidationUtils.java:
```diff
@@ -0,0 +1,40 @@
+        Period period = Period.between(currentDate, recommendation.getCreatedAt().toLocalDate());
+
+        if (period.toTotalMonths() <= 6) {
+            throw new DateTimeException("It is not possible to give a recommendationDto to this user, since the last " +
```
Кажется, что в ошибке лучше напистаь про recommendation

SergeyErmoshenko commented on src/main/java/school/faang/user_service/utils/ValidationUtils.java:
```diff
@@ -0,0 +1,40 @@
+        List<SkillOffer> skillOffersList = skillOfferRepository.findAllSkillOffers();
+        List<SkillOffer> skillOffersIds = recommendation.getSkillOffers();
+
+        if (!new HashSet<>(skillOffersList).containsAll(skillOffersIds)) {
```
Есил skillOffersList - null, то будет NPE

SergeyErmoshenko commented on src/main/java/school/faang/user_service/utils/ValidationUtils.java:
```diff
@@ -0,0 +1,40 @@
+        List<SkillOffer> skillOffersIds = recommendation.getSkillOffers();
+
+        if (!new HashSet<>(skillOffersList).containsAll(skillOffersIds)) {
+            throw new DataValidationException("All recommendationDto skills must be in the system.");
```
В сообщении лучше написать про Recommendation

SergeyErmoshenko commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,63 @@
+    private void addGuarantee(Recommendation recommendation) {
+        List<SkillOffer> skillOfferListOfReceiver =
+                skillOfferRepository.findAllByUserId(recommendation.getReceiver().getId());
+        for (SkillOffer soOfRecommendation : recommendation.getSkillOffers()) {
```
Название soOfRecommendation лучше поменять

SergeyErmoshenko commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,63 @@
+                UserSkillGuarantee userSkillGuarantee = new UserSkillGuarantee();
+                userSkillGuarantee.setUser(recommendation.getReceiver());
+                userSkillGuarantee.setSkill(soOfRecommendation.getSkill());
+                userSkillGuarantee.setGuarantor(recommendation.getAuthor());
```
Может userSkillGuarantee записать в БД?

SergeyErmoshenko commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,63 @@
+        createInSkillOfferRepository(recommendation);
+        addGuarantee(recommendation);
+        createInRecommendationRepository(recommendation);
+        recommendation.getSkillOffers().forEach(skillOfferRepository::save);
```
Ты вызываешь на List<SkillOffer> у recommendation сначала skillOfferRepository.create, а потом skillOfferRepository::save. Возможно стоит оставить create и не использовать базовый функционал CRUD

SergeyErmoshenko commented on src/main/java/school/faang/user_service/utils/ValidationUtils.java:
```diff
@@ -0,0 +1,40 @@
+
+    public static void validateRecommendationDate(Recommendation recommendation) {
+        LocalDate currentDate = LocalDate.now();
+        Period period = Period.between(currentDate, recommendation.getCreatedAt().toLocalDate());
```
Если createdAt у recommendation - null, то будет NPE

SergeyErmoshenko commented on src/main/java/school/faang/user_service/mapper/RecommendationMapper.java:
```diff
@@ -0,0 +1,59 @@
+@Mapper(componentModel = "spring")
+public interface RecommendationMapper {
+    @Mapping(target = "author", source = "authorId", qualifiedByName = "mapAuthorIdToUser")
+    @Mapping(target = "receiver", source = "receiverId", qualifiedByName = "mapReceiverIdToUser")
```
Можно попробовать избавиться от qualifiedByName, заменив два метода маппинга на один mapIdToUser, который будет преобразовывать Long id в User

SergeyErmoshenko commented on src/main/java/school/faang/user_service/repository/recommendation/SkillOfferRepository.java:
```diff
@@ -33,4 +33,9 @@ SELECT COUNT(so.id) FROM skill_offer so
     List<SkillOffer> findAllByUserId(long userId);
+
+    @Query(nativeQuery = true, value = """
+            SELECT DISTINCT * FROM skill_offer
```
Может попробуем заменить этот запрос, используя findAll из CRUD?

SergeyErmoshenko commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,63 @@
+    private final SkillOfferRepository skillOfferRepository;
+    private final RecommendationMapper recommendationMapper;
+
+
```
Лишняя пустая строка

SergeyErmoshenko commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,63 @@
+
+    private void createInSkillOfferRepository(Recommendation recommendation) {
+        List<SkillOffer> skillOffersOfRecommendation = recommendation.getSkillOffers();
+        skillOffersOfRecommendation.forEach(skillOffer ->
```
Если skill - null, то NPE

SergeyErmoshenko commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,63 @@
+
+    private void createInRecommendationRepository(Recommendation recommendation) {
+        Long authorId = recommendation.getAuthor().getId();
+        Long receiverId = recommendation.getReceiver().getId();
```
Если receiver - null, то NPE

SergeyErmoshenko commented:
Так же еще убрать замену большого кол-ва импортов на *, добавить логи

Commit:
добавил: ValidationRecommendationUtils, SkillDto
исправил: RecommendationMapper, RecommendationController, RecommendationService,
 RecommendationDto,
 SkillOfferDto

Commit:
исправил импорты

Commit:
исправил SkillOfferRepository и RecommendationService

Commit:
исправил RecommendationMapper
удалил SkillDto.java

Commit:
убрал лишние аннотации

Commit:
-

Commit:
сохранил список гарантов

Commit:
исправил маппер и валидацию

PavelEzhkov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,79 @@
+    private final UserSkillGuaranteeRepository userSkillGuaranteeRepository;
+
+    public RecommendationDto create(RecommendationDto recommendationDto) {
+        Recommendation recommendation = recommendationMapper.toRecommendation(recommendationDto);
```
Валидации должны быть до того как ты сходил в бд, ходить в бд сильно дороже по ресурсам, возможно нам 
это и не надо

PavelEzhkov replied:
Я так и не понял зачем тебе этот маппинг, кажется что можно работать с дто

PavelEzhkov commented on src/main/java/school/faang/user_service/utils/ValidationRecommendationUtils.java:
```diff
@@ -0,0 +1,55 @@
+    private static final String SKILL_OFFER_VALID_EXCEPTION = "Предложение о приобретении навыка с идентификатором: " +
+            "%d недействительно.";
+
+    public static void validateRecommendationContent(RecommendationDto recommendationDto) {
```
Кажется что тут нет необходимости передавать всю дто

PavelEzhkov commented on src/main/java/school/faang/user_service/utils/ValidationRecommendationUtils.java:
```diff
@@ -0,0 +1,55 @@
+        }
+    }
+
+    public static void validateRecommendationDate(RecommendationDto recommendationDto) {
```
Кажется что тут нет необходимости передавать всю дто

PavelEzhkov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,79 @@
+
+    public RecommendationDto create(RecommendationDto recommendationDto) {
+        Recommendation recommendation = recommendationMapper.toRecommendation(recommendationDto);
+        Iterable<SkillOffer> skillOffersIterable = skillOfferRepository.findAll();
```
Предлагаю сделать в репе метод 
 `List<SkillOffer> findAll();`
 и оставить 1 строчку

PavelEzhkov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,79 @@
+        skillOffersIterable.forEach(allSkillOffers::add);
+
+        ValidationRecommendationUtils.validateRecommendationContent(recommendationDto);
+        ValidationRecommendationUtils.validateRecommendationDate(recommendationDto);
```
нужно проверить, что автор дает рекомендацию не раньше, чем через 6 месяцев после его последней рекомендации этому пользователю, а ты проверяешь просто дату

PavelEzhkov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,79 @@
+        return recommendationMapper.toRecommendationDto(recommendation);
+    }
+
+    private void createInSkillOfferRepository(Recommendation recommendation) {
```
Не самое хорошее название метода createSkillOffer мне кажется что хватит

PavelEzhkov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,79 @@
+    private void createInSkillOfferRepository(Recommendation recommendation) {
+        List<SkillOffer> skillOffersOfRecommendation = recommendation.getSkillOffers();
+        for (SkillOffer skillOffer : skillOffersOfRecommendation) {
+            if (skillOffer == null) {
```
Кажется что эта проверка не нужна, тк от бд не может прийти null в листе

PavelEzhkov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,79 @@
+            if (skillOffer == null) {
+                throw new DataValidationException("Skill offer can't be null");
+            }
+            skillOfferRepository.create(skillOffer.getSkill().getId(), recommendation.getId());
```
Из скилл офера можно достать скилл и добавить ему гаранта и кажется, что надо это делать тут, а не отдельно

PavelEzhkov replied:
У тебя нет recommendation.getId() еще на до в начале сохранить recommendationRepository.create(authorId, receiverId, content); и там получить id

Commit:
исправил RecommendationService(не до конца), SkillOfferRepository, ValidationRecommendationUtils

Commit:
исправил RecommendationService,
RecommendationValidation
добавил метод в RecommendationRepository

Commit:
исправил RecommendationRepository

Commit:
Испаврил RecommendationMapper
удалил UserDto

PavelEzhkov commented on src/main/java/school/faang/user_service/mapper/RecommendationMapper.java:
```diff
@@ -0,0 +1,57 @@
+    @Mapping(target = "authorId", source = "author.id")
+    @Mapping(target = "receiverId", source = "receiver.id")
+    @Mapping(target = "skillOffersDto", source = "skillOffers")
+    RecommendationDto toRecommendationDto(Recommendation recommendation);
```
Рекомендации и скилы должны быть в разных маперах

PavelEzhkov commented on src/main/java/school/faang/user_service/repository/recommendation/RecommendationRepository.java:
```diff
@@ -29,4 +30,10 @@ INSERT INTO recommendation (author_id, receiver_id, content)
 
     Optional<Recommendation> findFirstByAuthorIdAndReceiverIdOrderByCreatedAtDesc(long authorId, long receiverId);
+
+    default LocalDateTime findLastRecommendationDate(long authorId, long receiverId) {
```
Не надо так делать) либо используй метод findFirstByAuthorIdAndReceiverIdOrderByCreatedAtDesc и делай поиск в сервисе либо пиши запрос к бд который сразу вернет что надо

PavelEzhkov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,72 @@
+        RecommendationValidation.validateRecommendationDate(lastRecommendation);
+
+        Recommendation recommendation = recommendationMapper.toRecommendation(recommendationDto);
+        recommendation.setCreatedAt(LocalDateTime.now());
```
так в дто есть уже сreatedAt

PavelEzhkov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,72 @@
+        for (SkillOffer skillOffer : skillOffersOfRecommendation) {
+            skillOfferRepository.create(skillOffer.getSkill().getId(), recommendation.getId());
+            if (skillOfferListOfReceiver.contains(skillOffer)) {
+                UserSkillGuarantee userSkillGuarantee = new UserSkillGuarantee();
```
Я бы предпочел использовать билдер тут
